### PR TITLE
Only capture kernel callchains and use general mechanism for perf for user space

### PR DESCRIPTION
**Pull request by luhenry**
This removes the need for allocating pinned memory to capture user space
stacktraces. We can now use perf_event_open only as a sampling
mechanism.

### CODE REVIEW

File: src/perfEvents_linux.cpp [hunk 1]
```diff
@@ -569,13 +579,16 @@ int PerfEvents::registerThread(int tid) {
...
 
-    void* page = mmap(NULL, 2 * OS::page_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
-    if (page == MAP_FAILED) {
-        Log::warn("perf_event mmap failed: %s", strerror(errno));
-        page = NULL;
+    void* page = NULL;
+    if ((_ring & RING_KERNEL)) {
+        page = mmap(NULL, 2 * OS::page_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
+        if (page == MAP_FAILED) {
+            Log::info("perf_event mmap failed: %s", strerror(errno));
```
Comment (jbachorik) [inline src/perfEvents_linux.cpp:589]: Should not this still be a warning?
Or is this kind of expected to happen quite frequently?

File: src/perfEvents_linux.cpp [hunk 2]
```diff
@@ -743,34 +757,71 @@ Error PerfEvents::start(Arguments& args) {
...
-    }
+    // As we don't have snapshot feature, it's wasteful to unregister all the threads
+    // to re-register them right after when doing a stop+start to capture the data.
+    // Instead, since we know we are continuously profiling and we know the interval
+    // doesn't change, simply don't unregister threads on stop, and check whether the
+    // thread has been registered already on start.
+
+    // for (int tid = 0; tid < _max_events; tid++) {
+    //     unregisterThread(tid);
+    // }
```
Comment (jbachorik) [inline src/perfEvents_linux.cpp:821]: This could be perhaps removed completely.

**Review by jbachorik — APPROVED** [review]
Just a few clarifying questions

File: src/perfEvents_linux.cpp (hunk 1 above)
Comment (luhenry) [reply src/perfEvents_linux.cpp:589]: It a fairly arcane setting on the Linux Kernel, and it only impacts capturing kernel stack traces. Having it `warn` would mean that we can get quite spammed on "normal" execution as it happens for _all_ threads.

File: src/perfEvents_linux.cpp (hunk 2 above)
Comment (luhenry) [reply src/perfEvents_linux.cpp:821]: I just left it for reference to clarify the context of the above comment. Happy to remove it if you think it's better.

Comment (jbachorik) [reply src/perfEvents_linux.cpp:821]: No strong opinion here. Usually, I just leave it to git to preserve the history ..

**Pull request merged**